vllm/core/block_manager.py

```diff
@@ -15,40 +63,87 @@ class BlockAllocator:
     the reference count becomes zero, the block is added back to the free list.
     """
 
-    def __init__(
-        self,
-        device: Device,
-        block_size: int,
-        num_blocks: int,
-    ) -> None:
+    def __init__(self,
+                 device: Device,
+                 block_size: int,
+                 num_blocks: int,
+                 eviction_policy: EvictionPolicy = EvictionPolicy.LRU) -> None:
         self.device = device
         self.block_size = block_size
         self.num_blocks = num_blocks
+        self.eviction_policy = eviction_policy
+
+        self.current_num_blocks = 0
+        self.table: Dict[int, PhysicalTokenBlock] = {}
+        self.free_table: Dict[int, PhysicalTokenBlock] = {}
 
-        # Initialize the free blocks.
-        self.free_blocks: BlockTable = []
-        for i in range(num_blocks):
-            block = PhysicalTokenBlock(device=device,
-                                       block_number=i,
-                                       block_size=block_size)
-            self.free_blocks.append(block)
-
-    def allocate(self) -> PhysicalTokenBlock:
-        if not self.free_blocks:
-            raise ValueError("Out of memory! No free blocks are available.")
-        block = self.free_blocks.pop()
-        block.ref_count = 1
+        self.default_hash_ctr = count()
+
+    def evict(self) -> PhysicalTokenBlock:
+        if self.eviction_policy == EvictionPolicy.LRU:
+            return lru_eviction(self.free_table)
+        else:
+            raise ValueError(
+                f"Unknown cache eviction policy: {self.eviction_policy}")
```

Comment:
I refactored the eviction logic a bit so that there's now an abstract Evictor class with LRU and FIFO subclasses that manage the data structure used to store the free blocks. With that refactoring, I've moved this dispatching logic to a `make_evictor(policy)` function which seems like a more natural place for it. LMK what you think.